components/encryption/src/file_dict_file.rs

```diff
@@ -280,10 +288,11 @@ impl FileDictionaryFile {
 
     fn parse_next_record(mut remained: &[u8]) -> Result<(usize, String, LogRecord)> {
         if remained.len() < Self::RECORD_HEADER_SIZE {
-            return Err(box_err!(
-                "file corrupted! record header size is too small: {}",
-                remained.len()
-            ));
+            warn!(
```

Comment:
It is better for ParseIncomplete to return this information then to just log it. If all the information is returned with the error then there is no need for a logging statement here: the error will end up getting logged higher up. Proper modeling would be to make an enum out of ParseIncomplete, but you could also have a struct of all the fields or even but everything into a single String.

Reply:
though I think it is fine to print the warning log here, where we have more context for detailed info.

Reply:
yeah, we can attach an error message with the error type. though contextual info, such as the remaining size here, can be printed as warning log in the current place.

Reply:
and this is not an actual error. This PR change to tolerate the tail corruption. we are using the special error type the denote the situation.

Reply:
File corruption seems like an error. I think you are saying that the error is recoverable. This function though should not know that: recovery is implemented by some caller. It might know that the error type indicates that it is recoverable, but it doesn't know if the callers properly implement recovery.
Logging here is also a case of knowing about the behavior of callers (that this doesn't need to be returned to a client or that it will be recovered).
Doing errors properly can certainly be heavyweight, so you will see me take shortcuts all the time, but I try to recognize that and differentiate it from the ideal error handling scenario.

Reply:
Yeah, we will say in the error message that it is the tail record being corrupted, and the caller need know no more than that. Contextual info like corruption offset etc should stay here IMO.

Reply:
Contextual information must be added here, but why must it stay here? In Go we use errors.Annotate to good effect (this would be similar to anyhow, which I need to try out). We end up with a caller at some point logging the error, and when it is logged it has the full context from multiple function calls.

Reply:
I know `errors.Annotate`, and I don't think it is a must. I want this piece of code print the warning log for easy debugging, where the info is only relevant here. I don't want to assume the caller will print the error.

Reply:
@gregwebs patterns are great, but I also prefer simplicity.

Reply:
@yiwu-arbug Why are you worried that the caller won't print the error? That would imply that a caller would just ignore an error: have you encountered this behavior in TiKV? Or are you concerned that the caller will not print it correctly?
I agree that logging immediately is much easier given the existing Rust tools. I will investigate different approaches.